src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedEntryPointSymbol.cs

```diff
@@ -329,6 +336,135 @@ private static BoundCall CreateParameterlessCall(CSharpSyntaxNode syntax, BoundE
             { WasCompilerGenerated = true };
         }
 
+        /// <summary> A synthesized entrypoint that forwards all calls to an async Main Method </summary>
+        internal sealed class AsyncForwardEntryPoint : SynthesizedEntryPointSymbol
+        {
+            /// The user-defined asynchronous main method.
+            private readonly MethodSymbol _userMain;
+
+            private readonly MethodSymbol _getAwaiterMethod;
+            private readonly MethodSymbol _getResultMethod;
+
+            private readonly ImmutableArray<ParameterSymbol> _parameters;
+
+            /// <summary>
+            /// Task -> Void
+            /// Generic Task -> Int32
+            /// </summary>
+            private static TypeSymbol TranslateReturnType(CSharpCompilation compilation, TypeSymbol returnType, DiagnosticBag diagnosticBag, MethodSymbol userMain)
+            {
+                var specialType = returnType.IsGenericTaskType(compilation) ? SpecialType.System_Int32 : SpecialType.System_Void;
+                var syntax = ExtractReturnTypeSyntax(userMain.DeclaringSyntaxReferences);
+                return Binder.GetSpecialType(compilation, specialType, syntax, diagnosticBag);
+            }
+
+            private static SyntaxNode ExtractReturnTypeSyntax(IEnumerable<SyntaxReference> userMainDeclaringReferences) {
+                var methodSyntax = userMainDeclaringReferences
+                    .Select(r => r.GetSyntax() as MethodDeclarationSyntax)
```

Comment:
>.Select(r => r.GetSyntax() as MethodDeclarationSyntax) [](start = 20, length = 54)

Linq is banned in compiler's code, all this can easily be done in a loop.